blues/simulation.py

```diff
@@ -364,78 +372,84 @@ def acceptRejectNCMC(self):
+            if write_move:
+                self.writeFrame(self.md_sim, 'acc-it%s.pdb' %(self.current_iter))
         else:
             self.reject += 1
-            print('NCMC MOVE REJECTED: log_ncmc {} < {}'.format(log_ncmc, randnum) )
+            self.logger.info('NCMC MOVE REJECTED: log_ncmc {} < {}'.format(log_ncmc, randnum) )
             self.nc_context.setPositions(md_state0['positions'])
 
         self.nc_integrator.reset()
-        self.md_sim.context.setVelocitiesToTemperature(self.temperature)
+        self.md_sim.context.setVelocitiesToTemperature(temperature)
 
-    def simulateNCMC(self, verbose=False, write_ncmc=False):
+    def simulateNCMC(self, nstepsNC=5000, nprop=5, ncmc_traj=None,
+                    reporter_interval=1000, **opt):
         """Function that performs the NCMC simulation."""
-        #append nc reporter at the first step
-        if (self.current_iter == 0) and (write_ncmc):
-            self.ncmc_reporter = app.dcdreporter.DCDReporter(self.ncmc_outfile, 1)
-            self.nc_sim.reporters.append(self.ncmc_reporter)
+
+        self.logger.info('[Iter %i] Starting NCMC Simulation with %i NCMC steps...' % (self.current_iter, nstepsNC))
+
         #choose a move to be performed according to move probabilities
         #TODO: will have to change to work with multiple alch regions
         self.move_engine.selectMove()
-        for nc_step in range(self.nstepsNC):
+        for nc_step in range(int(nstepsNC)):
             try:
-                self.current_stepNC = int(nc_step)
-                # Calculate Work/Energies Before Step
-                if verbose:
-                    work_initial = self.getWorkInfo(self.nc_integrator, self.work_keys)
-
                 # Attempt selected MoveEngine Move at the halfway point
                 #to ensure protocol is symmetric
-                if self.nstepsNC / 2 == nc_step:
-
+                if self.movestep == nc_step:
                     #Do move
-                    print('[Iter {}] Performing NCMC move'.format(self.current_iter))
+                    self.logger.info('Step = %s Performing NCMC move...' % nc_step)
                     self.nc_context = self.move_engine.runEngine(self.nc_context)
 
-                    if write_ncmc and (nc_step+1) % write_ncmc == 0:
-                        self.ncmc_reporter.report(self.nc_sim, self.nc_sim.context.getState(getPositions=True, getVelocities=True))
-
-                # Do 1 NCMC step with the integrator, alchemically modifying system
+                # Add extra propagation steps
+                if nprop > 1:
+                    current_lambda = self.nc_integrator.getGlobalVariableByName('lambda')
+                    if current_lambda >= 0.2 and current_lambda <= 0.8:
+                        for n in range(int(nprop)):
+                            self.nc_integrator.setGlobalVariableByName('lambda', self.nc_integrator.getGlobalVariableByName('lambda') - self.d_lambda)
```

Comment:
Besides the above, I think the way you handle adding propagation steps is fine although I'm not 100% sure. During those steps we're still accumulating protocol work from changing lambda and I think the integrator should correctly account for the change back in lambda when you reset it, but again I'm not confidently sure. 

I think the easiest way to check this (and maybe something that could be a good test to have) is to set all the masses of the system to zero in the NCMC simulation and perform an iteration using some non-1 nprop. Because the positions aren't changing the protocol work for that iteration should be essentially zero (you're turning off/on the ligand in the exact same place).